### ISSUE TITLE
Loconet Throttle with ESP32

### ISSUE THREAD

**Issue by jakespeed1311** [mrrwa/LocoNet2#2]
Hello,
i want to setup a Throttle with an ESP32.  but in the Init phase the CPU looks up.
I habve an NodeMUC ESP-32s wiht an ESP-Wroom32 Chip. Some other Tests with TFT ans so an are ok.

I have the Arduino IDE with the last ESP 1.0.1 installed. (Updated after the problem beginns)

For me its crash at       Throttle.init (0,0,0,  ThrottleID);      //erzeuge Handregler
I don't know whats wrong. 
Any change to help ?
Tank you very mutch for a hint.
Jake

`
#include <LocoNetESP32.h>

  uint16_t ThrottleID = 0x20;
  #define SpeedMax 127     //höchste Geschwindigkeit. (vielleicht auch hier Stufen nutzen -> FredI!)
 long unsigned int              LastThrottleTimerTick;

word Adresse = 1;      // default Adresse 16-bit
byte Slot = 0xFF;      // akt. Slot zur Adr. (kein)
byte AdrH = 0;         // high Anteil Adresse
byte AdrL = 0;         // low Anteil Adresse

LocoNetESP32Class LocoNet;

LocoNetThrottleClass Throttle;

void setup(void)
{
  Serial.begin(115200);
  Serial.println("Setup Start");

//### Loconet Init

     Serial.println("Init");
     LocoNet.init();                          // First initialize the LocoNet interface
     Serial.println("Throttle");
  
     Throttle.init (0,0,0,  ThrottleID);      //erzeuge Handregler
   
    Serial.println("Throttle OK");

  Serial.println("Setup END");
}

//--------------------------------------------------------------------------------------------------
void loop() { 
  //AdrL = Adresse;        //Adresse in je 2x 8 Bit Speichen für EEPROM Sicherung
  //AdrH = Adresse >> 8;
}

`

10:45:34.806 -> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
10:45:34.806 -> configsip: 0, SPIWP:0xee
10:45:34.806 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
10:45:34.806 -> mode:DIO, clock div:1
10:45:34.806 -> load:0x3fff0018,len:4
10:45:34.806 -> load:0x3fff001c,len:1100
10:45:34.806 -> load:0x40078000,len:10088
10:45:34.806 -> load:0x40080400,len:6380
10:45:34.806 -> entry 0x400806a4
10:45:34.966 -> Setup Start
10:45:34.966 -> Init
10:45:34.966 -> LocoNetESP32Class Initialising
10:45:34.966 -> Throttle
10:45:34.966 -> Guru Meditation Error: Core  1 pan⸮ets Jun  8 2016 00:22:57
⸮ets Jun  8 2016 00:22:57
10:45:35.046 -> 
10:45:35.046 -> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
10:45:35.046 -> configsip: 0, SPIWP:0xee
10:45:35.046 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
10:45:35.046 -> mode:DIO, clock div:1
10:45:35.046 -> load:0x3fff0018,len:4
10:45:35.046 -> load:0x3fff001c,len:1100
10:45:35.046 -> load:0x40078000,len:10088
10:45:35.046 -> load:0x40080400,len:6380
10:45:35.046 -> entry 0x400806a4
10:45:35.206 -> Setup Start
10:45:35.206 -> Init
10:45:35.206 -> LocoNetESP32Class Initialising
10:45:35.206 -> Throttle
10:45:35.206 -> Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
10:45:35.206 -> Core 1 register dump:
10:45:35.206 -> PC      : 0x400e9f37  PS      : 0x00060130  A0      : 0x800d0c5c  A1      : 0x3ffb1f60  
10:45:35.206 -> A2      : 0x3ffbfcf0  A3      : 0x00000000  A4      : 0x00000000  A5      : 0x00000000  
10:45:35.206 -> A6      : 0x00000020  A7      : 0x00000000  A8      : 0x800d1d30  A9      : 0x3ffb1f40  
10:45:35.246 -> A10     : 0x00000002  A11     : 0x3f400031  A12     : 0x00000001  A13     : 0x00000000  
10:45:35.246 -> A14     : 0xe03c1000  A15     : 0x3ff5f000  SAR     : 0x0000001c  EXCCAUSE: 0x0000001c  
10:45:35.246 -> EXCVADDR: 0x00000030  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xffffffff  
10:45:35.246 -> 
10:45:35.246 -> Backtrace: 0x400e9f37:0x3ffb1f60 0x400d0c59:0x3ffb1f80 0x400d2abb:0x3ffb1fb0 0x40089a09:0x3ffb1fd0
10:45:35.246 -> 
10:45:35.246 -> Rebooting...
10:45:35.286 -> ets Jun  8 2016 00:22:57
10:45:35.286 -> 
10:45:35.286 -> rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
10:45:35.286 -> configsip: 0, SPIWP:0xee
10:45:35.286 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
10:45:35.286 -> mode:DIO, clock div:1
10:45:35.286 -> load:0x3fff0018,len:4
10:45:35.286 -> load:0x3fff001c,len:1100
10:45:35.286 -> load:0x40078000,len:10088
10:45:35.286 -> load:0x40080400,len:6380
10:45:35.286 -> entry 0x400806a4
10:45:35.434 -> Setup Start

**Comment by mcmadhatter** [comment]
I haven't yet tested the Throttle functionality, but the first parameter of Throttle.init  is a pointer.  Could you try
 `Throttle.init (&LocoNet,0,0,  ThrottleID);`

**Comment by kiwi64ajs** [comment]
Hi Guys,

> On 6/02/2019, at 11:12 AM, mcmadhatter <notifications@github.com> wrote:
> 
> I haven't yet tested the Throttle functionality, but the first parameter of Throttle.init is a pointer. Could you try
> Throttle.init (LocoNet,0,0, ThrottleID);
> 
The init() method has the following signature:

	void init(LocoNetClass * lnInstance, uint8_t UserData, uint8_t Options, uint16_t ThrottleId);

You must pass in a valid pointer to a LocoNetClass derived class.

LocoNet2 evolved from my original, very AVR specific LocoNet library and I tried to do it in a much more modular way but still hide the complexity of how it all hangs together.

I refactored the generic LocoNet modules (Throttle, FastClock, SVs, LNCVs) to require a pointer to the core LocoNetClass base class and add themselves into a Linked-List in the LocoNetClass, so they can get notified when relevant LocoNet messages are received.

Its been many years since I looked at this code and there are subsequent bug-fixes in the original LocoNet library that should really be carried over to this code as well… Need to go back and review the diffs since the fork and figure out what needs to be changed.

HTH

Regards

Alex Shepherd

p: +64-7-829 4441
m: +64-21-777764
e: kiwi64ajs@gmail.com
